test/distributed/_tensor/test_dtensor.py

```diff
@@ -704,5 +707,265 @@ def test_split_tensor(self) -> None:
+        y_ref = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        ref = fn(x_ref, y_ref)
+
+        x = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        y = torch.arange(8, requires_grad=True, dtype=torch.float32)
+        res = opt_fn(x, y)
+
+        self.assertEqual(res, ref)
+
+        # Now run and assert the backward + gradients
+        ref.sum().backward()
+        res.sum().backward()
+
+        # The bw graph here is pretty messy (there are unnecessary cat and split calls?)
+        # But the important ting to note is:
+        # (1) we have a c10d_functional op in the graph, thanks to DTensor
+        # (2) we have a wait() call in the graph that inductor can reorder
+        # (3) We have a mm() in the graph (from the user code)
+        self.assertExpectedInline(
+            bw_graph_cell[0].code.strip(),
+            """\
+def forward(self, t_1, clone, tangents_1):
+    mm_3 = torch.ops.aten.mm.default(t_1, tangents_1);  t_1 = None
+    all_gather_into_tensor_2 = torch.ops.c10d_functional.all_gather_into_tensor.default(clone, 'ptd:0', [0, 1], 2);  clone = None
+    wait_tensor_2 = torch.ops.c10d_functional.wait_tensor.default(all_gather_into_tensor_2);  all_gather_into_tensor_2 = None
+    mm_5 = torch.ops.aten.mm.default(tangents_1, wait_tensor_2);  tangents_1 = wait_tensor_2 = None
+    split_1 = torch.ops.aten.split.Tensor(mm_3, 2, 1);  mm_3 = None
+    getitem_2 = split_1[0]
+    getitem_3 = split_1[1];  split_1 = None
+    cat_1 = torch.ops.aten.cat.default([getitem_2, getitem_3]);  getitem_2 = getitem_3 = None
+    reduce_scatter_tensor = torch.ops.c10d_functional.reduce_scatter_tensor.default(cat_1, 'SUM', 'ptd:0', [0, 1], 2);  cat_1 = None
+    wait_tensor_3 = torch.ops.c10d_functional.wait_tensor.default(reduce_scatter_tensor);  reduce_scatter_tensor = None
+    view_5 = torch.ops.aten.view.default(wait_tensor_3, [4, 2]);  wait_tensor_3 = None
+    _to_copy_2 = torch.ops.aten._to_copy.default(view_5, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'));  view_5 = None
+    view_6 = torch.ops.aten.view.default(_to_copy_2, [8]);  _to_copy_2 = None
+    view_7 = torch.ops.aten.view.default(mm_5, [2, 4]);  mm_5 = None
+    _to_copy_3 = torch.ops.aten._to_copy.default(view_7, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'));  view_7 = None
+    view_8 = torch.ops.aten.view.default(_to_copy_3, [8]);  _to_copy_3 = None
+    return [view_8, view_6]""",  # noqa: B950
+        )
+
+        # TODO: unfortunately we are getting incorrect gradients :(
+        # Ideally, we should be able to stare at the backward graph above and figure out what we're computing incorrectly.
+        # y_ref.grad = tensor([0., 1., 2., 3., 4., 5., 6., 7.])
+        # y.grad     = tensor([4., 4., 4., 4., 6., 6., 6., 6.])
+
+        self.assertEqual(x_ref.grad, x.grad)
+        # self.assertEqual(y_ref.grad, y.grad)
```

Comment:
darn, @wanchaol looks like we're computing an incorrect gradient in the backward in this test. I know you looked into a flavor of this problem a month or two ago- what was your fix again? I'll poke around, but if you have any ideas let me know!

Reply:
It's because of fake_pg, we can't rely on data correctness when using fake_pg, so if we want to test out the gradient correctness, we need to use multithreaded pg probably

Reply:
wow haha- I knew this and forgot it.